File: repertoire/concepts/data-integration/providers/capture_mode/FileUpload.swift
// Data Integration Kit - File Upload Capture Provider
// Direct file ingestion with MIME detection via magic bytes and metadata extraction

import Foundation

public final class FileUploadCaptureProvider {

    private struct MagicSignature {
        let bytes: [UInt8]
        let offset: Int
        let mimeType: String
        let fileExtension: String
    }

    private static let signatures: [MagicSignature] = [
        MagicSignature(bytes: [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A], offset: 0, mimeType: "image/png", fileExtension: "png"),
        MagicSignature(bytes: [0xFF, 0xD8, 0xFF], offset: 0, mimeType: "image/jpeg", fileExtension: "jpg"),
        MagicSignature(bytes: [0x47, 0x49, 0x46, 0x38, 0x37, 0x61], offset: 0, mimeType: "image/gif", fileExtension: "gif"),
        MagicSignature(bytes: [0x47, 0x49, 0x46, 0x38, 0x39, 0x61], offset: 0, mimeType: "image/gif", fileExtension: "gif"),
        MagicSignature(bytes: [0x25, 0x50, 0x44, 0x46], offset: 0, mimeType: "application/pdf", fileExtension: "pdf"),
        MagicSignature(bytes: [0x50, 0x4B, 0x03, 0x04], offset: 0, mimeType: "application/zip", fileExtension: "zip"),
        MagicSignature(bytes: [0x52, 0x49, 0x46, 0x46], offset: 0, mimeType: "image/webp", fileExtension: "webp"),
        MagicSignature(bytes: [0x42, 0x4D], offset: 0, mimeType: "image/bmp", fileExtension: "bmp"),
        MagicSignature(bytes: [0x49, 0x44, 0x33], offset: 0, mimeType: "audio/mpeg", fileExtension: "mp3"),
        MagicSignature(bytes: [0x66, 0x4C, 0x61, 0x43], offset: 0, mimeType: "audio/flac", fileExtension: "flac"),
        MagicSignature(bytes: [0x4F, 0x67, 0x67, 0x53], offset: 0, mimeType: "audio/ogg", fileExtension: "ogg"),
        MagicSignature(bytes: [0x1A, 0x45, 0xDF, 0xA3], offset: 0, mimeType: "video/webm", fileExtension: "webm"),
    ]

    public init() {}

    public func capture(input: CaptureInput, config: CaptureConfig) async throws -> CaptureItem {
        guard let fileData = input.file, !fileData.isEmpty else {
            throw CaptureError.parseError("file_upload capture requires a file buffer")
        }

        let (mimeType, ext) = detectMimeType(data: fileData)
        let fileSize = fileData.count
        let filename = (config.options?["filename"] as? String) ?? "upload.\(ext)"

        let dimensions: (width: Int, height: Int)?
        if mimeType.hasPrefix("image/") {
            dimensions = extractImageDimensions(data: fileData, mimeType: mimeType)
        } else {
            dimensions = nil
        }

        var summaryParts = [
            "File: \(filename)",
            "Type: \(mimeType)",
            "Size: \(formatFileSize(bytes: fileSize))"
        ]
        if let dims = dimensions {
            summaryParts.append("Dimensions: \(dims.width)x\(dims.height)")
        }

        let isText = mimeType.hasPrefix("text/") || mimeType == "application/json"
        let content: String
        if isText && fileSize < 1_048_576 {
            content = String(data: fileData, encoding: .utf8) ?? summaryParts.joined(separator: "\n")
        } else {
            content = summaryParts.joined(separator: "\n")
        }

        var tags = [ext, String(mimeType.split(separator: "/").first ?? "file")]
        if let dims = dimensions {
            tags.append("\(dims.width)x\(dims.height)")
        }

        return CaptureItem(
            content: content,
            sourceMetadata: SourceMetadata(
                title: filename,
                url: nil,
                capturedAt: ISO8601DateFormatter().string(from: Date()),
                contentType: mimeType,
                author: nil,
                tags: tags,
                source: "file_upload"
            ),
            rawData: (config.options?["includeBuffer"] as? Bool == true)
                ? ["data": fileData, "dimensions": dimensions as Any, "detectedMime": mimeType] as [String: Any]
                : nil
        )
    }

    public func supports(input: CaptureInput) -> Bool {
        guard let file = input.file else { return false }
        return !file.isEmpty
    }

    // MARK: - MIME Detection via Magic Bytes

    private func detectMimeType(data: Data) -> (mimeType: String, extension: String) {
        let bytes = [UInt8](data)
        for sig in Self.signatures {
            if bytes.count < sig.offset + sig.bytes.count { continue }
            let slice = Array(bytes[sig.offset..<sig.offset + sig.bytes.count])
            if slice == sig.bytes {
                return (sig.mimeType, sig.fileExtension)
            }
        }

        // Check text-based formats
        if let header = String(data: data.prefix(512), encoding: .utf8) {
            let trimmed = header.trimmingCharacters(in: .whitespaces)
            if trimmed.hasPrefix("<?xml") || trimmed.hasPrefix("<svg") {
                return ("image/svg+xml", "svg")
            }
            if trimmed.hasPrefix("<!DOCTYPE html") || trimmed.hasPrefix("<html") {
                return ("text/html", "html")
            }
            if trimmed.hasPrefix("{") || trimmed.hasPrefix("[") {
                return ("application/json", "json")
            }
        }
        if String(data: data, encoding: .utf8) != nil {
            return ("text/plain", "txt")
        }
        return ("application/octet-stream", "bin")
    }

    // MARK: - Image Dimension Extraction

    private func extractImageDimensions(data: Data, mimeType: String) -> (width: Int, height: Int)? {
        let bytes = [UInt8](data)
        switch mimeType {
        case "image/png":
            return extractPngDimensions(bytes: bytes)
        case "image/jpeg":
            return extractJpegDimensions(bytes: bytes)
        case "image/gif":
            return extractGifDimensions(bytes: bytes)
        default:
            return nil
        }
    }

    private func extractPngDimensions(bytes: [UInt8]) -> (width: Int, height: Int)? {
        guard bytes.count >= 24, bytes[0] == 0x89, bytes[1] == 0x50 else { return nil }
        let width = Int(UInt32(bytes[16]) << 24 | UInt32(bytes[17]) << 16 | UInt32(bytes[18]) << 8 | UInt32(bytes[19]))
        let height = Int(UInt32(bytes[20]) << 24 | UInt32(bytes[21]) << 16 | UInt32(bytes[22]) << 8 | UInt32(bytes[23]))
        return (width, height)
    }

    private func extractJpegDimensions(bytes: [UInt8]) -> (width: Int, height: Int)? {
        var offset = 2
        while offset + 8 < bytes.count {
            guard bytes[offset] == 0xFF else { offset += 1; continue }
            let marker = bytes[offset + 1]
            if marker == 0xC0 || marker == 0xC2 {
                let height = Int(UInt16(bytes[offset + 5]) << 8 | UInt16(bytes[offset + 6]))
                let width = Int(UInt16(bytes[offset + 7]) << 8 | UInt16(bytes[offset + 8]))
                return (width, height)
            }
            let segLen = Int(UInt16(bytes[offset + 2]) << 8 | UInt16(bytes[offset + 3]))
            offset += 2 + segLen
        }
        return nil
    }

    private func extractGifDimensions(bytes: [UInt8]) -> (width: Int, height: Int)? {
        guard bytes.count >= 10 else { return nil }
        let width = Int(UInt16(bytes[6]) | UInt16(bytes[7]) << 8)
        let height = Int(UInt16(bytes[8]) | UInt16(bytes[9]) << 8)
        return (width, height)
    }

    // MARK: - Utilities

    private func formatFileSize(bytes: Int) -> String {
        if bytes < 1024 { return "\(bytes) B" }
        if bytes < 1_048_576 { return String(format: "%.1f KB", Double(bytes) / 1024.0) }
        if bytes < 1_073_741_824 { return String(format: "%.1f MB", Double(bytes) / 1_048_576.0) }
        return String(format: "%.1f GB", Double(bytes) / 1_073_741_824.0)
    }
}

File: repertoire/concepts/data-integration/providers/capture_mode/file_upload.rs
// Data Integration Kit - File Upload Capture Provider
// Direct file ingestion with MIME detection via magic bytes and metadata extraction

use std::collections::HashMap;

pub const PROVIDER_ID: &str = "file_upload";
pub const PLUGIN_TYPE: &str = "capture_mode";

#[derive(Debug, Clone)]
pub struct CaptureInput {
    pub url: Option<String>,
    pub file: Option<Vec<u8>>,
    pub email: Option<String>,
    pub share_data: Option<serde_json::Value>,
}

#[derive(Debug, Clone)]
pub struct CaptureConfig {
    pub mode: String,
    pub options: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct SourceMetadata {
    pub title: String,
    pub url: Option<String>,
    pub captured_at: String,
    pub content_type: String,
    pub author: Option<String>,
    pub tags: Option<Vec<String>>,
    pub source: Option<String>,
}

#[derive(Debug, Clone)]
pub struct CaptureItem {
    pub content: String,
    pub source_metadata: SourceMetadata,
    pub raw_data: Option<Vec<u8>>,
}

#[derive(Debug)]
pub enum CaptureError {
    MissingFile,
    InvalidFile(String),
}

impl std::fmt::Display for CaptureError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CaptureError::MissingFile => write!(f, "file_upload capture requires a file buffer"),
            CaptureError::InvalidFile(e) => write!(f, "Invalid file: {}", e),
        }
    }
}

struct MagicSignature {
    bytes: &'static [u8],
    offset: usize,
    mime_type: &'static str,
    extension: &'static str,
}

const MAGIC_SIGNATURES: &[MagicSignature] = &[
    MagicSignature { bytes: &[0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A], offset: 0, mime_type: "image/png", extension: "png" },
    MagicSignature { bytes: &[0xFF, 0xD8, 0xFF], offset: 0, mime_type: "image/jpeg", extension: "jpg" },
    MagicSignature { bytes: &[0x47, 0x49, 0x46, 0x38, 0x37, 0x61], offset: 0, mime_type: "image/gif", extension: "gif" },
    MagicSignature { bytes: &[0x47, 0x49, 0x46, 0x38, 0x39, 0x61], offset: 0, mime_type: "image/gif", extension: "gif" },
    MagicSignature { bytes: &[0x25, 0x50, 0x44, 0x46], offset: 0, mime_type: "application/pdf", extension: "pdf" },
    MagicSignature { bytes: &[0x50, 0x4B, 0x03, 0x04], offset: 0, mime_type: "application/zip", extension: "zip" },
    MagicSignature { bytes: &[0x52, 0x49, 0x46, 0x46], offset: 0, mime_type: "image/webp", extension: "webp" },
    MagicSignature { bytes: &[0x42, 0x4D], offset: 0, mime_type: "image/bmp", extension: "bmp" },
    MagicSignature { bytes: &[0x49, 0x44, 0x33], offset: 0, mime_type: "audio/mpeg", extension: "mp3" },
    MagicSignature { bytes: &[0x66, 0x4C, 0x61, 0x43], offset: 0, mime_type: "audio/flac", extension: "flac" },
    MagicSignature { bytes: &[0x4F, 0x67, 0x67, 0x53], offset: 0, mime_type: "audio/ogg", extension: "ogg" },
    MagicSignature { bytes: &[0x1A, 0x45, 0xDF, 0xA3], offset: 0, mime_type: "video/webm", extension: "webm" },
];

#[derive(Debug, Clone)]
pub struct ImageDimensions {
    pub width: u32,
    pub height: u32,
}

fn detect_mime_type(data: &[u8]) -> (&'static str, &'static str) {
    for sig in MAGIC_SIGNATURES {
        if data.len() < sig.offset + sig.bytes.len() { continue; }
        let slice = &data[sig.offset..sig.offset + sig.bytes.len()];
        if slice == sig.bytes {
            return (sig.mime_type, sig.extension);
        }
    }

    // Check text-based formats
    let header = std::str::from_utf8(&data[..std::cmp::min(data.len(), 512)]).unwrap_or("");
    let trimmed = header.trim_start();
    if trimmed.starts_with("<?xml") || trimmed.starts_with("<svg") {
        return ("image/svg+xml", "svg");
    }
    if trimmed.starts_with("<!DOCTYPE html") || trimmed.starts_with("<html") {
        return ("text/html", "html");
    }
    if trimmed.starts_with('{') || trimmed.starts_with('[') {
        return ("application/json", "json");
    }
    // Check if valid UTF-8 text
    if std::str::from_utf8(data).is_ok() {
        return ("text/plain", "txt");
    }
    ("application/octet-stream", "bin")
}

fn extract_png_dimensions(data: &[u8]) -> Option<ImageDimensions> {
    if data.len() < 24 { return None; }
    if data[0] != 0x89 || data[1] != 0x50 { return None; }
    let width = u32::from_be_bytes([data[16], data[17], data[18], data[19]]);
    let height = u32::from_be_bytes([data[20], data[21], data[22], data[23]]);
    Some(ImageDimensions { width, height })
}

fn extract_jpeg_dimensions(data: &[u8]) -> Option<ImageDimensions> {
    let mut offset = 2usize;
    while offset + 8 < data.len() {
        if data[offset] != 0xFF { offset += 1; continue; }
        let marker = data[offset + 1];
        if marker == 0xC0 || marker == 0xC2 {
            let height = u16::from_be_bytes([data[offset + 5], data[offset + 6]]) as u32;
            let width = u16::from_be_bytes([data[offset + 7], data[offset + 8]]) as u32;
            return Some(ImageDimensions { width, height });
        }
        let seg_len = u16::from_be_bytes([data[offset + 2], data[offset + 3]]) as usize;
        offset += 2 + seg_len;
    }
    None
}

fn extract_gif_dimensions(data: &[u8]) -> Option<ImageDimensions> {
    if data.len() < 10 { return None; }
    let width = u16::from_le_bytes([data[6], data[7]]) as u32;
    let height = u16::from_le_bytes([data[8], data[9]]) as u32;
    Some(ImageDimensions { width, height })
}

fn extract_image_dimensions(data: &[u8], mime_type: &str) -> Option<ImageDimensions> {
    match mime_type {
        "image/png" => extract_png_dimensions(data),
        "image/jpeg" => extract_jpeg_dimensions(data),
        "image/gif" => extract_gif_dimensions(data),
        _ => None,
    }
}

fn format_file_size(bytes: usize) -> String {
    if bytes < 1024 { return format!("{} B", bytes); }
    if bytes < 1048576 { return format!("{:.1} KB", bytes as f64 / 1024.0); }
    if bytes < 1073741824 { return format!("{:.1} MB", bytes as f64 / 1048576.0); }
    format!("{:.1} GB", bytes as f64 / 1073741824.0)
}

pub struct FileUploadCaptureProvider;

impl FileUploadCaptureProvider {
    pub fn new() -> Self { Self }

    pub fn capture(&self, input: &CaptureInput, config: &CaptureConfig) -> Result<CaptureItem, CaptureError> {
        let file_data = input.file.as_ref().ok_or(CaptureError::MissingFile)?;
        if file_data.is_empty() { return Err(CaptureError::MissingFile); }

        let (mime_type, extension) = detect_mime_type(file_data);
        let file_size = file_data.len();
        let filename = config.options.as_ref()
            .and_then(|o| o.get("filename"))
            .and_then(|v| v.as_str())
            .map(|s| s.to_string())
            .unwrap_or_else(|| format!("upload.{}", extension));

        let dimensions = if mime_type.starts_with("image/") {
            extract_image_dimensions(file_data, mime_type)
        } else {
            None
        };

        let mut summary = vec![
            format!("File: {}", filename),
            format!("Type: {}", mime_type),
            format!("Size: {}", format_file_size(file_size)),
        ];
        if let Some(ref dims) = dimensions {
            summary.push(format!("Dimensions: {}x{}", dims.width, dims.height));
        }

        let is_text = mime_type.starts_with("text/") || mime_type == "application/json";
        let content = if is_text && file_size < 1048576 {
            String::from_utf8_lossy(file_data).to_string()
        } else {
            summary.join("\n")
        };

        let mut tags = vec![extension.to_string(), mime_type.split('/').next().unwrap_or("file").to_string()];
        if let Some(ref dims) = dimensions {
            tags.push(format!("{}x{}", dims.width, dims.height));
        }

        Ok(CaptureItem {
            content,
            source_metadata: SourceMetadata {
                title: filename,
                url: None,
                captured_at: chrono::Utc::now().to_rfc3339(),
                content_type: mime_type.to_string(),
                author: None,
                tags: Some(tags),
                source: Some("file_upload".to_string()),
            },
            raw_data: None,
        })
    }

    pub fn supports(&self, input: &CaptureInput) -> bool {
        input.file.as_ref().map_or(false, |f| !f.is_empty())
    }
}
